### ISSUE TITLE
How to save a trained model to a specific local path?

### ISSUE THREAD

**Issue by WizardandCat** [microsoft/qlib#636]
## ❓ Questions and Help

How can I save an trained model to a specific local path? I have tried this but failed:

`with R.start(experiment_name='test_1'):
        R.save_objects(**{"model.pkl": model}, local_path='./trained_results/model', artifact_path='./trained_results')`

It just remind me that "No tracking URI is provided. Use the default tracking URI." And then the model and the params are saved to the default "mlruns" fold.

By the way, after I have saved the model to a specific local path, how can I load it the next time? Is it right to code like this:

`recorder = R.get_recorder(recorder_id=rid, experiment_name="test_1")
        model = recorder.load_object("./trained_results/model.pkl")`

(Should I still use "test_1" as the "experiment_name"? )

Thank you!

**Comment by you-n-g** [comment]
@WizardandCat 
I add more detailed docs for the API.
Can your problem be solved with them? 

https://github.com/microsoft/qlib/pull/639/files

**Comment by WizardandCat** [comment]
> @WizardandCat I add more detailed docs for the API. Can your problem be solved with them?
> 
> https://github.com/microsoft/qlib/pull/639/files


Thanks for your help! But I still failed to run the code in your updated case. This is the problem I encountered:

I have create these variables:
``` python
task = 
{'model':` {'class': 'LGBModel',
  'module_path': 'qlib.contrib.model.gbdt',
  'kwargs': {...},
 'dataset': {'class': 'DatasetH',
  'module_path': 'qlib.data.dataset',
  'kwargs': {'handler': {'class': 'Alpha158',
    'module_path': 'qlib.contrib.data.handler',
    'kwargs': {'start_time': '2017-01-01',
     'end_time': '2021-09-30',
     'fit_start_time': '2017-01-01',
     'fit_end_time': '2020-12-31',
     'instruments': {'market': 'all',
      'filter_pipe': [...]}}},
   'segments': {'train': ('2017-01-01', '2019-12-31'),
    'valid': ('2020-01-01', '2020-12-31'),
    'test': ('2021-01-01', '2021-09-30')}}}}

model = init_instance_by_config(task["model"])
dataset = init_instance_by_config(task["dataset"])
```


Then, I created **a new empty folder named “result”** in the same directory as this program. I hope to run the following code to train and save the model, the parameters and other data qlib need to load the next time I want to use this model **to a specific local path** './results/':
``` python
with R.start(experiment_name="test"):
    R.log_params(**flatten_dict(task))
    model.fit(dataset)
    R.save_objects(local_path='results/model.pkl', artifact_path="results")
    rid = R.get_recorder().id
```
But it just raise Error: No such file or directory: 'results/model.pkl'.
What should I do to fix this error?


Besides, when I write like this: 
``` python
R.save_objects(**{"pred.pkl": pred, "model.pkl": model}, artifact_path='results')
```
It success but it save all the data to the default fold "./mlruns/". But I hope to save them to the specific local path.


In addition, I also notice that someone write like this:
```
R.save_objects(trained_model=model)
```
But I search the "recorder.py" function and didn't find the variable "trained_model". How does it works?


At last, I'm really confused with: What's the relationship between "local_path" and "artifact_path"? What data are saved to the local_path and what are saved to the artifact_path?


Thanks for your time and your kindly help.

**Comment by you-n-g** [comment]
> It success but it save all the data to the default fold "./mlruns/". But I hope to save them to the specific local path.

When you use `R.save_objects`,  Qlib will save the objects you passed in to its experiment management backend (e.g. mlflow, the `mlruns` folder).
Such an experiment management system will provide you some APIs to make it easier to load previous model (e.g. `R.get_recorder(recorder_id=rid).load_object("prediction/pred.pkl")`)

If you want save your objects to specific `Path` instead of leveraging Qlib's experiment management backend,  you can use `pickle` to dump and load them directly.

> ut I search the "recorder.py" function and didn't find the variable "trained_model". How does it work?

It works by leveraging Python's [kwargs mechanism](https://book.pythontips.com/en/latest/args_and_kwargs.html).
Here is the related code https://github.com/microsoft/qlib/blob/main/qlib/workflow/recorder.py#L300

> At last, I'm really confused with: What's the relationship between "local_path" and "artifact_path"? What data are saved to the local_path and what are saved to the artifact_path?

You can refer to [mlflow's docs](https://mlflow.org/docs/0.4.2/python_api/mlflow.html#mlflow.log_artifacts) for the meaning of `local_path` and `artifact_path` 


In summary, this API is designs for saving **objects** to **the experiments management backend path**,
1. Qlib provide two methods to specify **objects**
- Passing in the object directly by passing with `**kwargs` (e.g. R.save_objects(trained_model=model))
- Passing in the local path to the object, i.e. `local_path` parameter.
2. `artifact_path` represents the  **the experiments management backend path**

**Comment by specialuse** [comment]
I meet the same problem. Saving mlruns at working directory will bring some troubles especially when the file is big.

**Comment by specialuse** [comment]
> ## ❓ Questions and Help
> How can I save an trained model to a specific local path? I have tried this but failed:
> 
> `with R.start(experiment_name='test_1'): R.save_objects(**{"model.pkl": model}, local_path='./trained_results/model', artifact_path='./trained_results')`
> 
> It just remind me that "No tracking URI is provided. Use the default tracking URI." And then the model and the params are saved to the default "mlruns" fold.
> 
> By the way, after I have saved the model to a specific local path, how can I load it the next time? Is it right to code like this:
> 
> `recorder = R.get_recorder(recorder_id=rid, experiment_name="test_1") model = recorder.load_object("./trained_results/model.pkl")`
> 
> (Should I still use "test_1" as the "experiment_name"? )
> 
> Thank you!



> ## ❓ Questions and Help
> How can I save an trained model to a specific local path? I have tried this but failed:
> 
> `with R.start(experiment_name='test_1'): R.save_objects(**{"model.pkl": model}, local_path='./trained_results/model', artifact_path='./trained_results')`
> 
> It just remind me that "No tracking URI is provided. Use the default tracking URI." And then the model and the params are saved to the default "mlruns" fold.
> 
> By the way, after I have saved the model to a specific local path, how can I load it the next time? Is it right to code like this:
> 
> `recorder = R.get_recorder(recorder_id=rid, experiment_name="test_1") model = recorder.load_object("./trained_results/model.pkl")`
> 
> (Should I still use "test_1" as the "experiment_name"? )
> 
> Thank you!

hi,have you saved the problems?
